lib/fluent/plugin/out_http.rb

```diff
@@ -57,34 +109,79 @@ def shutdown
     super
   end
 
+  def format_url(tag, time, record)
+    @endpoint_url
+  end
+
   def set_body(req, tag, time, record)
     if @serializer == :json
       set_json_body(req, record)
+    elsif @serializer == :text
+      set_text_body(req, record)
+    elsif @serializer == :raw
+      set_raw_body(req, record)
     else
       req.set_form_data(record)
     end
     req
   end
 
   def set_header(req, tag, time, record)
-    req
+    if @custom_headers
+      @custom_headers.each do |k,v|
+        req[k] = v
+      end
+      req
+    else
+      req
+    end
   end
 
   def set_json_body(req, data)
     req.body = Yajl.dump(data)
     req['Content-Type'] = 'application/json'
   end
 
+  def set_text_body(req, data)
+    req.body = data["message"]
+    req['Content-Type'] = 'text/plain'
+  end
+
+  def set_raw_body(req, data)
+    req.body = data.to_s
+    req['Content-Type'] = 'application/octet-stream'
+  end
+
+  def create_request(tag, time, record)
+    url = format_url(tag, time, record)
+    uri = URI.parse(url)
+    req = Net::HTTP.const_get(@http_method.to_s.capitalize).new(uri.request_uri)
+    set_body(req, tag, time, record)
+    set_header(req, tag, time, record)
+    return req, uri
+  end
+
   def http_opts(uri)
       opts = {
         :use_ssl => uri.scheme == 'https'
       }
       opts[:verify_mode] = @ssl_verify_mode if opts[:use_ssl]
+      opts[:ca_file] = File.join(@ca_file) if File.file?(@ca_file)
       opts
   end
 
+  def proxies
+    ENV['HTTPS_PROXY'] || ENV['HTTP_PROXY'] || ENV['http_proxy'] || ENV['https_proxy']
+  end
+
   def send_request(record)
-    res = @adapter.post(@uri.path) do |request|
+    is_rate_limited = (@rate_limit_msec != 0 and not @last_request_time.nil?)
```

Comment:
Readability vs. speed, I think assigning variables in Ruby takes extra cycles to allocate memory space